File: Data Structures & Algorithms/meeting-schedule-ii/submission-1.cpp
/**
 * Definition of Interval:
 * class Interval {
 * public:
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 * }
 */

class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        if(intervals.size() == 0)
            return 0;
        vector<int> v(1000010, 0);
        int ans = 0;
        for(int i = 0; i < intervals.size() ; i++){
            v[intervals[i].start]++;
            v[intervals[i].end]--;
        }

        for(int i = 1 ; i < 1000010 ; i++){
            v[i] += v[i-1];
            ans = max(ans, v[i]);
        }
        return ans;
    }
};

File: Data Structures & Algorithms/insert-new-interval/submission-1.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int len = intervals.length;
        List<int[]> l = new ArrayList<>();
        for(int i = 0 ; i < len ; i++){
            l.add(intervals[i]);
        }
        l.add(newInterval);
        Collections.sort(l, (a,b)-> a[0]-b[0]);
        List<int[]> ans = new ArrayList<>();
        int currentStart = l.get(0)[0];
        int currentEnd = l.get(0)[1];
        for(int i = 1 ; i <= len ; i++){
            if(l.get(i)[0] > currentEnd){
                ans.add(new int[]{currentStart, currentEnd});
                currentStart = l.get(i)[0];
                currentEnd = l.get(i)[1];
            } else {
                currentEnd = Math.max(l.get(i)[1], currentEnd);
            }
        } 

        ans.add(new int[]{currentStart, currentEnd});  
        int[][] finalAns = new int[ans.size()][];
        int i = 0;
        for(int[] interval : ans){
            finalAns[i++] = interval;
        }
        return finalAns; 
    }
}
